### PR TITLE
[WEB-4025] Add useTheming hook, reduce generated TW classes

### PR DESCRIPTION
**Pull request by jamiehenson**
## Jira Ticket Link / Motivation

There's been a few swings at getting dynamic theming right across our components, but so far things haven't been quite right in terms of optimisation and the usage of the functionality in the code - this PR aims to address things on both those fronts.

## Summary of changes

Currently, the approach to generating extra Tailwind classes necessitated by dynamic theming has embedded a misunderstanding about Tailwind itself. Turns out, very large lookup tables are ingested by Tailwind wholesale, and result in very large build sizes.

This PR flips the approach by scanning the repo for uses of the `themeColor` helper (previously `t`), and keeping a record of the thematic variants of those classes _only_, meaning we only generate the classes we actually need instead of every possible combination.

This is supported by the restructuring of the dynamic theming logic into a hook called `useTheming`. The advantages of this are twofold:
- it standardises what the theming helper is called. Previously, `t` was a convention but this wasn't reinforced anywhere, which is messy, particularly when you need some consistency when scanning for theme-oriented class names in the code. `useTheming` only gives you `themeColor` back.
- it paves the way for global theming in future, as we can integrate this hook in with an repo-wide context which holds the state for theming across a whole app

## How do you manually test this?

It's a background change so hard to test directly. Look at Storybook and see all the colours work on theme-flipped components (i.e. PricingCards) I guess.

## Merge/Deploy Checklist

- [ ] Written automated tests for implemented features/fixed bugs
- [ ] Rebased and squashed commits
- [ ] Commits have clear descriptions of their changes
- [ ] Checked for any performance regressions

## Frontend Checklist

- [ ] No frontend changes in this PR
- [ ] Added before/after screenshots for changes
- [ ] Tested on different platforms/browsers with [Browserstack](https://www.browserstack.com)
- [ ] Compared with the initial design / our [brand guidelines](https://www.figma.com/file/jTQgyIovrhGBOf4Zrvjvbk/Ably-Linear-Design?node-id=118%3A0)
- [ ] Checked the code for accessibility issues ([VoiceOver User Guide](https://support.apple.com/en-gb/guide/voiceover/welcome/mac))?

## Summary by CodeRabbit

- **New Features**
	- Introduced a new theming hook (`useTheming`) to manage theme colors dynamically across components.
	- Enhanced color management with new definitions and structures for color types.

- **Bug Fixes**
	- Removed deprecated `determineThemeColor` function to streamline theme color management.

- **Documentation**
	- Updated documentation for `DynamicTheming` to reflect new usage patterns with the `useTheming` hook.

- **Chores**
	- Incremented version number in `package.json` to indicate a shift to a development version.
	- Modified `.gitignore` to adjust tracking behavior for specific color files.

### CODE REVIEW

File: src/core/hooks/useTheming.tsx [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+      theme === baseTheme ? color : invertTailwindClassVariant(color),
+    [baseTheme, theme],
+  );
+
+  return {
+    themeColor,
+  };
+};
+
+export default useTheming;
```
[the remark this answers is not in the document]
Comment (jamiehenson) [reply src/core/hooks/useTheming.tsx:25]: Not required here, if I need further flexibility then I will restructure the hook at a higher level

File: scripts/compute-colors.ts [hunk 2]
```diff
@@ -1,72 +1,69 @@
...
+  const readDirectory = (dir: string) => {
+    const files = fs.readdirSync(dir);
+
+    files.forEach((file) => {
+      const filePath = path.join(dir, file);
+      const stat = fs.statSync(filePath);
+
+      if (stat.isDirectory()) {
+        readDirectory(filePath);
+      } else if (filePath.endsWith(".tsx")) {
```
[the remark this answers is not in the document]
Comment (jamiehenson) [reply scripts/compute-colors.ts:37]: Not needed

File: scripts/compute-colors.ts [hunk 3]
```diff
@@ -1,72 +1,69 @@
...
+        const content = fs.readFileSync(filePath, "utf-8");
+        const regex = new RegExp(
+          `themeColor\\("((${joinedVariants}${joinedPrefixes})-(${colors})-(000|[1-9]00|1[0-3]00))"\\)`,
+          "g",
+        );
```
[the remark this answers is not in the document]
Comment (jamiehenson) [reply scripts/compute-colors.ts:42]: This is intentional

File: src/core/ProductTile.tsx [hunk 4]
```diff
@@ -20,14 +19,15 @@
...
+  });
   const { icon, label, description, link, unavailable } = products[name] ?? {};
 
-  const t = (color: ColorClass) =>
-    determineThemeColor("dark", selected ? "light" : "dark", color);
-
   return (
     <div
-      className={`rounded-lg p-12 flex flex-col gap-8 transition-colors ${t("bg-neutral-1200")} ${selected ? "" : "hover:bg-neutral-1100"} ${className ?? ""}`}
+      className={`rounded-lg p-12 flex flex-col gap-8 transition-colors ${themeColor("bg-neutral-1200")} ${selected ? "" : "hover:bg-neutral-1100"} ${className ?? ""}`}
```
[the remark this answers is not in the document]
Comment (jamiehenson) [reply src/core/ProductTile.tsx:30]: Nope

File: src/core/styles/colors/types.ts [hunk 5]
```diff
@@ -164,24 +176,19 @@ const aliasedColors = [
...
->;
+export const colorGroupLengths = {
+  neutral: neutralColors.length,
+  orange: orangeColors.length,
+  blue: blueColors.length,
+  yellow: yellowColors.length,
+  green: greenColors.length,
+  violet: violetColors.length,
+  pink: pinkColors.length,
+};
```
[the remark this answers is not in the document]
Comment (jamiehenson) [reply src/core/styles/colors/types.ts:194]: Not a problem

File: src/core/styles/colors/types.ts [hunk 6]
```diff
@@ -164,24 +176,19 @@ const aliasedColors = [
   "transparent",
 ] as const;
 
-export const colorNames = {
+export const colorRoles = {
   neutral: neutralColors,
   orange: orangeColors,
   secondary: secondaryColors,
   gui: guiColors,
```
[the remark this answers is not in the document]
Comment (jamiehenson) [reply src/core/styles/colors/types.ts:183]: But it hasn't been removed from `colorRoles`...

**Review by aralovelace — APPROVED** [review]
I just tested Pricing Cards in the storybook and all works.
I cant dig in to the logic yet but  I scanned it looks good as well. Cant wait for the next show and tell to know moore about this

File: scripts/compute-colors.ts [hunk 7]
```diff
@@ -1,72 +1,83 @@
...
+        return;
+      }
+
+      if (stat.isDirectory()) {
+        readDirectory(filePath);
+      } else if (filePath.endsWith(".tsx")) {
+        let content = "";
+        try {
+          content = fs.readFileSync(filePath, "utf-8");
+          const regex = new RegExp(
```
[the remark this answers is not in the document]
Comment (jamiehenson) [reply scripts/compute-colors.ts:45]: That's definitely a better way of doing it

**Comment by jamiehenson** [comment]
If the question though was, "how does this work in `voltaire`", then the generated json is injected into downstream repo Tailwind configs as before, there are just a lot fewer classes

File: scripts/compute-colors.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (jamiehenson) [reply scripts/compute-colors.ts:45]: I'll do this as I need to rebase this branch anyway with conflict resolutions

**Pull request merged**